src/main/commandServer/settingsValidator.ts

```diff
@@ -0,0 +1,185 @@
+          // It's an internal error (HTTP Status 500) if it isn't.
+          errors.push(`Setting ${ fqname } should be a simple value, but got <${ JSON.stringify(newSettings[k]) }>.`);
+        }
+      } else {
+        // Throw an exception if this field isn't a function, because in the verifier all values should be
+        // either child objects or functions.
+        changeNeeded = allowedSettings[k].call(this, currentSettings[k], newSettings[k], errors, fqname) || changeNeeded;
+        if (changeNeeded && fqname === 'kubernetes.containerEngine' && newSettings[k] === 'docker') {
+          newSettings[k] = 'moby';
+        }
+      }
+    }
+
+    return changeNeeded;
+  }
+
+  protected checkContainerEngine(currentValue: string, desiredEngine: string, errors: string[], fqname: string): boolean {
+    switch (desiredEngine) {
+    case 'containerd':
+    case 'moby':
+      break;
+    case 'docker':
+      desiredEngine = 'moby';
+      break;
+    default:
+      errors.push(`Invalid value for ${ fqname }: <${ desiredEngine }>; must be 'containerd', 'docker', or 'moby'`);
+
+      return false;
+    }
+
+    return currentValue !== desiredEngine;
+  }
+
+  protected checkEnabled(currentState: boolean, desiredState: string|boolean, errors: string[], fqname: string): boolean {
+    if (typeof (desiredState) !== 'boolean') {
+      switch (desiredState) {
+      case 'true':
+        desiredState = true;
+        break;
+      case 'false':
+        desiredState = false;
+        break;
+      default:
+        errors.push(`Invalid value for ${ fqname }: <${ desiredState }>`);
+
+        return false;
+      }
+    }
+
+    return currentState !== desiredState;
+  }
+
+  protected checkKubernetesVersion(currentValue: string, desiredVersion: string, errors: string[], fqname: string): boolean {
+    const ptn = /^v?(\d+\.\d+\.\d+)(?:\+k3s\d+)?$/;
+    const m = ptn.exec(desiredVersion);
+
+    if (!m) {
+      errors.push(`Desired kubernetes version not valid: <${ desiredVersion }>`);
+
+      return false;
+    }
+    desiredVersion = m[1];
+    if (this.k8sVersions.length === 0) {
+      errors.push(`Can't check field ${ fqname }: no versions of Kubernetes were found.`);
+
+      return false;
+    } else if (!this.k8sVersions.includes(desiredVersion)) {
+      errors.push(`Kubernetes version ${ desiredVersion } not found.`);
+
+      return false;
+    }
+
+    return currentValue !== desiredVersion;
+  }
+
+  protected notSupported(fqname: string) {
+    return `Changing field ${ fqname } via the API isn't supported.`;
+  }
+
+  protected checkUnchanged(currentValue: any, desiredValue: any, errors: string[], fqname: string): boolean {
+    if (currentValue !== desiredValue) {
+      errors.push(this.notSupported(fqname));
+    }
+
+    return false;
+  }
+
+  // only arrays support stringification, so convert objects to arrays of tuples and sort on the keys
+  protected stableSerialize(value: Record<string, any>) {
```

Comment:
This doesn't properly deal with nested children; e.g.
```ts
stableSerialize({
  root: {
    child: 1,
    childToo: 2,
    childMore: 3,
  }
});
```

Reply:
Not an issue because we won't compile that situation:

```
TS2322: Type '{ child: number; childToo: number; childMore: number; }' is not assignable to type 'boolean'.
```

I've changed variable names (and types) to make it clear that this check applies only to the `WSLIntegrations` -- they're currently the only pref field whose child takes arbitrary fields. If we add another such field in the future, it will probably needs its own checker.